environment.yml

```diff
@@ -0,0 +1,9 @@
+name: pudl
+channels:
+  - conda-forge
+dependencies:
+  - python=3.10
+  - pip
+  - pip:
+      - h3==3.7.6 # h3-py 3.7.4 on conda-forge fails, see https://github.com/scikit-build/scikit-build/pull/901
```

Comment:
If we'd like to pin all of our dependencies and rely on `conda-lock` to do that work for us, how do we need to change the way we think about creating our software environment?

The `conda-lock.yml` file gets checked into source control and is supposed to represent a set of platform-specific packages which exact versions that, if installed, is known to work (since we'll have run all our tests within that environment and gotten the ✅ right?

How would this affect our development environment? Right now we typically do development within the environment defined by `devtools/environment.yml`. Would we switch to doing development in the environment defined by `conda-lock.yml` plus some additional packages? In development we can only install `catalystcoop.pudl` using `pip -e` right now.

Reply:
How can we:
* avoid the need to specify our dependencies in more than one place.
* ensure that we have a fully specified environment that doesn't drift over time.
* periodically update all our dependencies intentionally and check that everything is still working (and if is, save the new set of dependencies as the default going forward).
* use the same environment for published packages, deployment (e.g. in CI, local tests with Tox, and our nightly builds) and also in the environment that we're developing in day to day?

Reply:
Thanks for the thoughtful review!

1. Short-term: Yes, I'll take a stab at branching off of current commit to solve apple silicon problem while we discuss this. (I totally missed `devtools/environment.yml` folder, sorry!)
2. Yes, you are totally correct that PUDL would move to using `conda-lock` as its source of truth for dependancies. So the test runners should boot up their environment using conda-lock, the prod system should run using conda-lock, etc. 
3. Development environment: I'm less puritanical here and I think the process would be installing the environment with conda-lock and then the package with `pip -e`. The important part is that _after_ you're done developing, you know that any tests and production will be done on the `conda-lock.yml` deps. So if your development work requires updating a package, or adding a new one, that has to be frozen into the lockfile. This will be enforced automatically when the test suite installs from the lockfile and then tries to run your code. 


I think one question you alluded to is "Do we need both an `environmental.yml` and a `conda-lock.yml`?!
I come from `Pipfile` world where every project has a `Pipfile` and a `Pipfile.lock` so I don't have a strong reaction to two files - one that's human readable, and another (lock file) that's really pinning down specific version dependancies. People get used to it pretty quickly. 

Now, using your bullet points as a ruberic:

Goal | Does conda-lock help?
---|----
Dependancies in only one place | Kind of. For python packages conda-lock is superior since it is pinning version numbers
Environment doesn't drift | Definitely
Periodically update dependancies and test that it works | Yes, definitely. This can be done by volunteers or an automated job runner that updates dependancies, sees if tests works, and then opens a PR. Updating deps _is_ a code change and should deserve the same rigor of testing
Use same env for published packages, deployment, and dev environment | I think so. Everyone should use the environment created by conda-lock and then `pip -e` for local development of the pudl code.